python/first_histogram.md

```diff
@@ -0,0 +1,242 @@
+# Making your first histogram
+
+In this section we are going to make our first plots using Python. For this we
+will use some public LHCb data for $$ B^+ \rightarrow H^+ H^+ H^- $$ from the
+[CERN open data portal](http://opendata.cern.ch/?ln=en). This data is available
+on EOS:
+
+```bash
+$ xrdfs root://eospublic.cern.ch/ ls /eos/opendata/lhcb/AntimatterMatters2017/data
+/eos/opendata/lhcb/AntimatterMatters2017/data/B2HHH_MagnetDown.root
+/eos/opendata/lhcb/AntimatterMatters2017/data/B2HHH_MagnetUp.root
+/eos/opendata/lhcb/AntimatterMatters2017/data/PhaseSpaceSimulation.root
+```
+
+{% callout "Installing python packages" %}
+One of the best things about Python is the huge number of packages available,
+for everything from art to machine learning to web development. While LCG
+provides a lot of useful packages there are sometimes things missing.
+Fortunately most packages can be easily installed into `~/.local/` using `pip`
+with the `--user` flag. In this lesson we will be using `root_pandas` which can
+be installed using:
+
+```bash
+pip install --user root_pandas
+```
+
+We can even upgrade already installed packages using:
+
+```bash
+pip install --user matplotlib --upgrade
+```
+
+When using LCG this requires that you also run
+```bash
+export PYTHONPATH=$HOME/.local/lib/python2.7/site-packages/:$PYTHONPATH
```

Comment:
Concerning `pip install --user` we've had some internal discussion on how to handle it. The problem is that the default installation directory is not necessarily `~/.local` - it changes depending on your Python distribution and your OS.

To force an installation directory for the `--user` option:

    export PYTHONUSERBASE=$HOME/.local
    export PATH=$PYTHONUSERBASE/bin:$PATH

`PYTHONUSERBASE` is a standard Python environment variable and does not require `PYTHONPATH` to be set accordingly. The `PATH` addition ensures that users can access Python executables installed by some Python packages.

The solution above works everywhere, in particular:

* Standard lxplus
* lxplus with a LCG view loaded (your `source /cvmfs...` line from above)
* macOS with standard Python
* macOS with Python from Homebrew

What do you think?

Reply:
`PYTHONPATH` isn't needed unless the package is also provided by LCG as `PYTHONPATH` has the highest priority in the search order so the `--user` installed package never gets used. This should ideally be fixed in LCG so they distribute `sitecustomize.py` or something rather than setting `PYTHONPATH`.

I'll try to explain more clearly that the `export` is only needed when upgrading LCG packages and modify it to use:

```
export PYTHONPATH=$(python -m site --user-site):$PYTHONPATH
```

> The PATH addition ensures that users can access Python executables installed by some Python packages.

This is a good point that I had forgotten, an alternative which doesn't require explicitly setting `PYTHONUSERBASE` is:

```bash
export PATH=$(python -m site --user-base)/bin:$PATH
```

What do you think?